bitcoin/src/blockdata/script/witness_program.rs

```diff
@@ -37,33 +41,43 @@ impl WitnessProgram {
         P: TryInto<PushBytesBuf>,
         <P as TryInto<PushBytesBuf>>::Error: PushBytesErrorReport,
     {
+        let program =
+            program.try_into().map_err(|error| Error::InvalidLength(error.input_len()))?;
+        WitnessProgram::from_bytes(version, program.as_bytes())
+    }
+
+    /// Creates a new witness program, copying the content from the given byte slice.
+    pub fn from_bytes(version: WitnessVersion, bytes: &[u8]) -> Result<Self, Error> {
         use Error::*;
 
-        let program = program.try_into().map_err(|error| InvalidLength(error.input_len()))?;
-        if program.len() < 2 || program.len() > 40 {
-            return Err(InvalidLength(program.len()));
+        let program_len = bytes.len();
+        if program_len < 2 || program_len > WITNESS_PROGRAM_MAX_SIZE {
+            return Err(InvalidLength(program_len));
```

Comment:
We could clear up the clippy warning and improve the error path by adding `Error::TooLong` and `Error::TooShort`
```rust
        if program_len < MIN_SIZE {
            return Err(TooShort(program_len));
        }

        if program_len > MAX_SIZE {
            return Err(TooLong(program_len));
```

Reply:
You do not have to use my suggestion, but if you like it here it is (since I hacked it up locally to see what it looks like anyways)
```rust
/// Witness program error.
#[derive(Clone, Debug, PartialEq, Eq)]
#[non_exhaustive]
pub enum Error {
    /// The witness program must be at least 2 bytes long.
    TooShort(usize),
    /// The witness program cannot be more than 40 bytes long.
    TooLong(usize),
    /// A v0 witness program must be either of length 20 or 32.
    InvalidSegwitV0Length(usize),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use Error::*;

        match *self {
            TooLong(ref len) => write!(f, "the witness program must be at least 2 bytes long: {}", len),
            TooShort(ref len) => write!(f, "the witness program cannot be more than 40 bytes long: {}", len),
            InvalidSegwitV0Length(len) => {
                write!(f, "a v0 witness program must be either 20 or 32 bytes: length={}", len)
            }
        }
    }
}

#[cfg(feature = "std")]
impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        use Error::*;

        match *self {
            TooShort(_) | TooLong(_) | InvalidSegwitV0Length(_) => None,
        }
    }
}
```

Reply:
Thanks, i'll keep `InvalidLength`. it seems weird to have different errors for too-short/too-long inputs. Just my preference